src/posix/platform/trel.cpp

```diff
@@ -0,0 +1,546 @@
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements platform for TREL using IPv6/UDP socket under POSIX.
+ */
+
+#include "openthread-posix-config.h"
+
+#include "platform-posix.h"
+
+#include <arpa/inet.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <net/if.h>
+#include <netinet/in.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include <openthread/platform/trel.h>
+
+#include "common/code_utils.hpp"
+#include "common/logging.hpp"
+
+#if OPENTHREAD_CONFIG_RADIO_LINK_TREL_ENABLE
+
+#define TREL_MAX_PACKET_SIZE 1400
+#define TREL_PACKET_POOL_SIZE 5
+
+typedef struct TxPacket
+{
+    struct TxPacket *mNext;
+    uint8_t          mBuffer[TREL_MAX_PACKET_SIZE];
+    uint16_t         mLength;
+    otSockAddr       mDestSockAddr;
+} TxPacket;
+
+static uint8_t   sRxPacketBuffer[TREL_MAX_PACKET_SIZE];
+static uint16_t  sRxPacketLength;
+static TxPacket  sTxPacketPool[TREL_PACKET_POOL_SIZE];
+static TxPacket *sFreeTxPacketHead;  // A singly linked list of free/available `TxPacket` from pool.
+static TxPacket *sTxPacketQueueTail; // A circular linked list for queued tx packets.
+
+static bool sInitialized = false;
+static bool sEnabled     = false;
+static int  sSocket      = -1;
+
+#if OPENTHREAD_CONFIG_LOG_PLATFORM && (OPENTHREAD_CONFIG_LOG_LEVEL >= OT_LOG_LEVEL_DEBG)
+
+static const char *Ip6AddrToString(const void *aAddress)
+{
+    static char string[INET6_ADDRSTRLEN];
+    return inet_ntop(AF_INET6, aAddress, string, sizeof(string));
+}
+
+static const char *BufferToString(const uint8_t *aBuffer, uint16_t aLength)
+{
+    const uint16_t kMaxWrite = 16;
+    static char    string[1600];
+
+    uint16_t num = 0;
+    char *   cur = &string[0];
+    char *   end = &string[sizeof(string) - 1];
+
+    cur += snprintf(cur, (uint16_t)(end - cur), "[(len:%d) ", aLength);
+    VerifyOrExit(cur < end);
+
+    while (aLength-- && (num < kMaxWrite))
+    {
+        cur += snprintf(cur, (uint16_t)(end - cur), "%02x ", *aBuffer++);
+        VerifyOrExit(cur < end);
+
+        num++;
+    }
+
+    if (aLength != 0)
+    {
+        cur += snprintf(cur, (uint16_t)(end - cur), "... ");
+        VerifyOrExit(cur < end);
+    }
+
+    *cur++ = ']';
+    VerifyOrExit(cur < end);
+
+    *cur++ = '\0';
+
+exit:
+    *end = '\0';
+    return string;
+}
+
+#endif // #if OPENTHREAD_CONFIG_LOG_PLATFORM && (OPENTHREAD_CONFIG_LOG_LEVEL >= OT_LOG_LEVEL_DEBG)
+
+static void PrepareSocket(uint16_t &aUdpPort)
+{
+    int                 val;
+    struct sockaddr_in6 sockAddr;
+    socklen_t           sockLen;
+
+    otLogDebgPlat("[trel] PrepareSocket()");
+
+    sSocket = SocketWithCloseExec(AF_INET6, SOCK_DGRAM, 0, kSocketNonBlock);
+    VerifyOrDie(sSocket >= 0, OT_EXIT_ERROR_ERRNO);
+
+    // Make the socket non-blocking to allow immediate tx attempt.
```

Comment:
Isn't the socket already non-blocking since we are using `kSocketNonBlock` ?

Reply:
yes. May be possible to remove/optimize  this, but I would prefer to keep it as is for now.
- This is not new code, it is more of less the exitsing code we had in `posix/trel_udp6.c` (I actually `git mv` this file to rename it, but for some reason on github it shows as if the old file is fully removed and new file is added).
- I don't think we are covering this code in any of the OT CI (we used this before issue 6472), so I feel uneasy changing things much (from what we know worked)